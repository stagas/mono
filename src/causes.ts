import { ParserErrorCauses } from 'tinypratt'

export class CompilerErrorCause extends ParserErrorCauses.ParserErrorCause {
  name = 'CompilerUnknownError'
}
export class ReferenceErrorCause extends CompilerErrorCause {
  name = 'CompilerReferenceError'
}
export class TypeErrorCause extends CompilerErrorCause {
  name = 'CompilerTypeError'
}
export class SyntaxErrorCause extends CompilerErrorCause {
  name = 'CompilerSyntaxError'
}
export class InvalidErrorCause extends CompilerErrorCause {
  name = 'CompilerInvalidError'
}
