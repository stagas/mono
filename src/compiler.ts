import { Typed, Type } from './typed'
import { Node, Token } from './parser'
import { SExpr, S } from './sexpr'
import { flatten, mush } from './util'

interface Op {
  (...sexprs: SExpr[]): SExpr
}

type CtxOp = TokenOp | NodeOp

interface TokenOp {
  (local: Context, ops: OpTable): RawTokenOp
}

interface NodeOp {
  (local: Context, ops: OpTable): RawNodeOp
}

interface RawTokenOp {
  (lhs: Token, ...nodes: Node[]): SExpr
}

interface RawNodeOp {
  (...nodes: Node[]): SExpr
}

interface OpTable {
  [k: string]: null | (() => CtxOp) | Op | Op[]
}

interface Scope {
  [k: string]: Type
}

interface Arg {
  id: Token
  default?: SExpr
  range?: SExpr
}

interface Context {
  scope: Scope
  args: Arg[]
}

type Func = [SExpr, SExpr]

export interface Imports {
  [k: string]: { params: Type[]; result: Type }
}

export interface Module {
  body: SExpr
  funcs: Record<string, Func>
  contexts: Map<Func, Context>
  valueOf(): SExpr
  toString(include?: string[]): string
}

export { Type }

export const compile = (node: Node, imports: Imports = {}) => {
  const global = { scope: {}, args: [] }
  const contexts: Module['contexts'] = new Map()
  const funcs: Module['funcs'] = {}

  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  const panic = (node as any).panic // TODO: resolve this in tinypratt

  // create types
  const { typeOf, typeAs, cast, castAll, hi, max, top, infer } = Typed(panic)

  for (const [name, { params, result }] of Object.entries(imports)) {
    const func: Func = [[], typeAs(result as Type, [])]
    funcs[name] = func
    const ctx: Context = {
      scope: {},
      args: params.map((x: string, i) => ({
        id: i.toString(),
        default: typeAs(x as Type, [x + '.const', '0']),
      })) as Arg[],
    }
    contexts.set(func, ctx)
  }

  /** todo is a "not implemented" marker for ops */
  const todo = null

  /** constructs a binary op of least type `type` */
  const bin =
    (type: Type, op: string): Op =>
    (lhs, rhs) =>
      top(max(type, hi(lhs, rhs)), [op, lhs, rhs])

  /** constructs a binary op of exact type `type` */
  const typebin =
    (type: Type, op: string): Op =>
    (lhs, rhs) =>
      top(type, [op, lhs, rhs])

  /** constructs an equality op */
  const eq =
    (op: string): Op =>
    (lhs, rhs) => {
      const type = max(Type.i32, hi(lhs, rhs))
      if (type === Type.f32) return typeAs(Type.bool, top(Type.f32, [op, lhs, rhs]))
      return typeAs(Type.bool, top(Type.i32, [op + '_s', lhs, rhs]))
    }

  /** defines a function */
  const funcDef = (ctx: Context, ops: OpTable, sym: string, args: Node[], rhs: Node) => {
    const body = map(flatten(';', rhs), ctx, ops)
    const func: Func = [map(args, ctx, OpArgs), typeAs(typeOf(body.at(-1)), body)]
    funcs[sym] = func
    contexts.set(func, ctx)
    return func
  }

  /** primary optable */
  const Op: OpTable = {
    ',': todo,
    ';': todo,
    '..': todo,

    '=': (): CtxOp => (local, ops) => (lhs, rhs) => {
      // f()=x : function declaration
      if (Array.isArray(lhs)) {
        if (lhs[0] != '@') throw new SyntaxError(panic('invalid assignment', lhs[0]))
        const [sym, args] = [lhs[1], flatten(',', lhs[2]).filter(Boolean)] as [Token, Node[]]
        const scope = Object.fromEntries(args.map(x => [x, Type.f32]))
        const ctx = { scope, args: [] }
        funcDef(ctx, ops, sym, args, rhs)
        return []
      }
      // x=y : variable assignment
      else {
        const symbol = lhs
        const value = build(rhs, local, ops)
        const scope = symbol in local.scope ? local.scope : symbol in global.scope ? global.scope : local.scope
        const type = symbol in scope ? scope[symbol] : (scope[symbol] = typeOf(value))
        return typeAs(type, [(scope === global.scope ? 'global' : 'local') + '.set', '$' + symbol, cast(type, value)])
      }
    },

    // x?y:z : ternary conditional
    '?': (cond, then_body, else_body) => {
      const type = hi(then_body, else_body)
      return typeAs(type, [
        'if',
        ['result', max(Type.i32, type)],
        cast(Type.bool, cond),
        ['then', cast(type, then_body)],
        ['else', cast(type, else_body)],
      ])
    },

    // logical Or
    '||':
      (): NodeOp =>
      (local, ops) =>
      (...nodes) => {
        const [lhs, rhs] = map(nodes, local, ops)
        const type = hi(lhs, rhs)
        const temp = '__lhs__' + type
        const zero = top(type, ['const', '0'])
        if (!(temp in local.scope)) local.scope[temp] = type
        return typeAs(type, [
          'if',
          ['result', max(Type.i32, type)],
          top(type, ['ne', zero, ['local.tee', temp, cast(type, lhs)]]),
          ['then', ['local.get', temp]],
          ['else', cast(type, rhs)],
        ])
      },

    // logical And
    // commented out because it's implemented as an AST rewrite
    // in parser as a ternary: lhs!=0?rhs:0, also it was wrong.
    // '&&': (lhs, rhs) => {
    //   const type = hi(lhs, rhs)
    //   const zero = top(type, ['const', '0'])
    //   return typeAs(type, [
    //     'if',
    //     ['result', max(Type.i32, type)],
    //     top(type, ['ne', zero, cast(type, lhs)]),
    //     ['then', cast(type, rhs)],
    //     ['else', zero],
    //   ])
    // },

    // x|y : bitwise OR
    '|': typebin(Type.i32, 'or'),

    // x^y : bitwise XOR
    '^': typebin(Type.i32, 'xor'),

    // x&y : bitwise AND
    '&': typebin(Type.i32, 'and'),

    '==': (lhs, rhs) => typeAs(Type.bool, bin(Type.i32, 'eq')(lhs, rhs)),
    '!=': (lhs, rhs) => typeAs(Type.bool, bin(Type.i32, 'ne')(lhs, rhs)),

    '<': eq('lt'),
    '>': eq('gt'),
    '<=': eq('le'),
    '>=': eq('ge'),

    // x>>y : bitwise shift right
    '>>': typebin(Type.i32, 'shr_s'),
    // x<<y : bitwise shift left
    '<<': typebin(Type.i32, 'shl'),

    '+': [
      // x+y : arithmetic add
      bin(Type.i32, 'add'),
      // +x  : cast to number
      x => cast(max(Type.i32, typeOf(x)), x),
    ],
    '-': [
      // x-y : arithmetic subtract
      bin(Type.i32, 'sub'),
      // -x  : arithmetic negate
      x => bin(Type.i32, 'mul')(top(max(Type.i32, typeOf(x)), ['const', '-1']), x),
    ],

    // x*y : arithmetic multiply
    '*': bin(Type.i32, 'mul'),
    // x/y : arithmetic divide
    '/': bin(Type.i32, 'div'),
    '%': (lhs, rhs) => {
      const type = hi(lhs, rhs)
      if (type === Type.f32) return ['call', '$mod', ...castAll(Type.f32, lhs, rhs)]
      if (type === Type.bool) return top(Type.i32, ['rem_s', lhs, rhs])
      return top(Type.i32, ['rem_u', lhs, rhs])
    },
    // !x : logical Not
    '!': x => top(Type.bool, ['eqz', x]),
    // ~x : bitwise NOT
    '~': x => top(Type.i32, ['not', x]),

    '[': todo,
    '(': todo,
    '@': (): CtxOp => (local, ops) => (sym, rhs) => {
      const func = funcs[sym]
      if (!func) throw new ReferenceError(panic('function not defined', sym))
      const body = func[1]
      const type = typeOf(body)
      const ctx = contexts.get(func)!
      // evaluate argument expressions
      const args = map(flatten(',', rhs), local, ops)
      // examine function argument declarations against passed arguments
      ctx.args.forEach((arg, i) => {
        // function argument declaration has default value
        if (arg.default) {
          // missing passed argument becomes the default value
          if (!args[i]) args[i] = arg.default
          // has passed argument but cast it to correct type
          else args[i] = cast(typeOf(arg.default), args[i])
        }
        // function argument declaration has range
        else if (arg.range) {
          // missing passed argument becomes the start of range value
          if (!args[i]) args[i] = arg.range[0] as SExpr
          // has passed argument but cast it to correct type
          else args[i] = cast(hi(...arg.range), args[i])
        }
        // has passed argument but no default, it is cast implicitly to f32
        else if (args[i]) args[i] = cast(Type.f32, args[i])
        // did not pass argument and no default, so implicitly push a zero f32 (0.0)
        else args[i] = ['f32.const', '0']
      })
      // truncate number of passed arguments down to the accepted function arguments
      args.length = ctx.args.length
      // call the function
      return typeAs(type, ['call', '$' + sym, ...args])
    },
    '.': todo,

    num: (): CtxOp => () => lit => top(infer(lit), ['const', lit]),

    ids: (): CtxOp => local => symbol => {
      const scope = symbol in local.scope ? local.scope : symbol in global.scope ? global.scope : local.scope
      if (!(symbol in scope)) throw new ReferenceError(panic('symbol not defined', symbol))
      const type = scope[symbol]
      return typeAs(type, [(scope === global.scope ? 'global' : 'local') + '.get', '$' + symbol])
    },
  }

  /** arguments optable */
  const OpArgs: OpTable = {
    ...Op,
    '..': (lhs, rhs) => [lhs, rhs],
    '=': (): CtxOp => (local, ops) => (id, value) => {
      // if it's not an atom then it has ranges
      if (Array.isArray(id)) id = build(id, local, ops)[0] as Token
      mush(local.args, { id, default: build(value, local, ops) })
      return [id]
    },
    '[': (): CtxOp => (local, ops) => (id, range) => (mush(local.args, { id, range: build(range, local, ops) }), [id]),
    ids: (): CtxOp => local => id => (mush(local.args, { id }), [id]),
  }

  /** builds a `node` under context `ctx` and optable `ops` */
  const build = (node: Node, ctx: Context, ops: OpTable): SExpr => {
    if (Array.isArray(node)) {
      const [sym, ...nodes] = node as [Token, Node[]]
      if (!sym || !nodes.length) return []
      let op = ops[sym]
      if (!op) throw new Error(panic('not implemented', sym))
      if (Array.isArray(op)) op = op.find(x => x.length === nodes.length) || op[0]
      return op.length ? (<Op>op)(...map(nodes, ctx, ops)) : (<() => NodeOp>op)()(ctx, ops)(...nodes)
    } else {
      const op = ops[node.group]
      if (!op) throw new Error(panic ? panic('not implemented', node) : 'not implemented: ' + node)
      return (<() => CtxOp>op)()(ctx, ops)(node)
    }
  }

  /** builds an array of `nodes` under context `ctx` and optable `ops` */
  const map = (nodes: Node[], ctx: Context, ops: OpTable): SExpr[] =>
    nodes
      .filter(Boolean)
      .map(x => build(x, ctx, ops))
      .filter(x => x.length > 0)

  // ==================================================================================
  // init

  // create start function
  funcDef(global, Op, '__start__', [], node)

  // create module
  const mod: Module = {
    body: [['start', '$__start__']] as SExpr,
    funcs,
    contexts,
    valueOf() {
      return this.body
    },
    toString(include = []) {
      return S(['module', ...include, ...this.body])
    },
  }

  // create functions
  for (const [sym, func] of Object.entries(funcs)) {
    if (sym in imports) continue
    const [args, body] = func
    const ctx = contexts.get(func)!
    mod.body.push([
      'func',
      '$' + sym,
      ['export', `"${sym}"`],
      ...args.map(x => ['param', '$' + x, ctx.scope[x as Token]]),
      ...(body.length && sym !== '__start__' ? [['result', max(Type.i32, typeOf(body))]] : []),
      ...Object.entries(ctx.scope)
        .filter(([x]) => !args.find(y => y == x))
        .map(([x, type]) => ['local', '$' + x, max(Type.i32, type)]),
      ...(sym === '__start__' ? (body.length ? [...body, 'drop'] : []) : body),
    ])
  }

  return mod
}
