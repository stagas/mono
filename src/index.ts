export * from './compiler'
export * from './get-shared-worker-port'
export * from './linker-service'
export * from './parser'
export * from './syntax'
