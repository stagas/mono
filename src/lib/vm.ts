import type { LinkerConfig } from '../linker'
import { Type, W, max } from '../typed'
import { Arg } from '..'

export const env = (config: LinkerConfig) => `;;wasm
  (import "env" "memory"
    (memory
      ${config.memory.initial}
      ${config.memory.maximum}
      ${config.memory.shared ? 'shared' : ''}
    )
  )

  (global $output_mem_ptr (i32) (i32.const 0))
  (global $input_mem_ptr (i32) (i32.const 0x40000))
  (global $global_mem_ptr (export "global_mem_ptr") (mut i32) (i32.const 0x80000))
  (global $sr (export "sampleRate") (mut f32) (f32.const 44100.0))
  (global $t (export "currentTime") (mut f32) (f32.const 0))
  (global $pi (f32) (f32.const 3.1415927410125732))
  (global $pi2 (f32) (f32.const 6.2831854820251465))
`

// TODO: length should be the global_buffer_length

export interface VM {
  fill(frame: number, length: number, ...args: number[]): void
}

export const fill = ({ type, params }: { type: Type; params: Arg[] }) => `;;wasm
  (func $fill (export "fill")
    ;; song position starting frame
    (param $frame i32)
    ;; buffer length to fill
    (param $length i32)
    ;; ...params
    ${params.map(x => `(param $${x.id} ${max(Type.i32, x.type!)})`).join(' ')}

    ;; buffer index position
    (local $index i32)

    ;; index = 0
    (local.set $index (i32.const 0))

    ;; do
    (loop $loop
      (global.set $global_mem_ptr (i32.const 0x80000))

      ;; run initializers
      (call $__begin__)

      ;; t = $frame / $sr
      (global.set $t
        (f32.div
          (f32.convert_i32_u (local.get $frame))
          (global.get $sr)
        )
      )

      ;; buffer[$index] =
      (f32.store (i32.shl (local.get $index) (i32.const 2))
        ;; f(...params)
        ${W(type) < W(Type.f32) ? '(f32.convert_i32_s ' : ''}
        (call $f ${params.map(x => `(local.get $${x.id})`).join(' ')})
        ${W(type) < W(Type.f32) ? ')' : ''}
      )

      ;; $index++
      (local.set $index (i32.add (local.get $index) (i32.const 1)))

      ;; $frame++
      (local.set $frame (i32.add (local.get $frame) (i32.const 1)))

      ;; if ($index !== $length) continue $loop
      (br_if $loop (i32.ne (local.get $index) (local.get $length)))
    )
  )
`
