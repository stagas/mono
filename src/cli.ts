import { arg, decarg } from 'decarg'
import * as fs from 'fs'
import * as path from 'path'
import watCompile from 'wat-compiler'

import { compile } from './compiler'
import { parse } from './parser'
import { S } from './sexpr'

const default_global = { global_mem_ptr: 'i32' }

function bodyOf(s: string, global: any = default_global) {
  return S(compile(parse(s), global).body)
}

export class Options {
  @arg('<file>', 'The .mono file to run') file!: string
  @arg('--', '[...args]', 'The arguments to pass to the function') args: (string | number)[] = []
  @arg('-n', '--name', 'The exported function name to use') name = 'f'
  @arg('-i', '--iterations', 'Bench iterations') iterations = 0

  static examples = {
    'my-app.mono': 'Runs my-app.mono',
    'my-app.mono -n add -- 1 2': 'Runs export `add` with arguments `(1,2)`',
    'my-app.mono -n add -i 100 -- 1 2': 'Runs the same 100 times and prints the result',
  }

  constructor(opts: Partial<Options> = {}) {
    Object.assign(this, opts)
  }

  get code() {
    return fs.readFileSync(path.resolve(process.cwd(), this.file), 'utf-8')
  }

  get parsedArgs() {
    return this.args.map(parseArgNumber)
  }
}

/**
 * Runs a function in a .wat file with the given arguments.
 */
export const run = (options: Options) => {
  const { name, code, parsedArgs } = options
  const instance = getInstance(code)
  const fn = instance.exports[name] as (...args: unknown[]) => unknown
  if (!fn) throw new Error(`Export "${name}" was not found.`)

  return fn(...parsedArgs)
}

export const bench = (options: Options) => {
  const { name, code, args } = options

  const instance = getInstance(`
    ${code}
    ${benchWat(
    name,
    args.map(x => `(${('' + x).includes('.') ? 'f32' : 'i32'}.const ${x})`).join(''),
    options.iterations
  )
    }
  `)

  const fn = instance.exports.bench as () => void

  console.time('bench')
  fn()
  console.timeEnd('bench')
}

export const getInstance = (code: string) => {
  console.log(code)
  const wat = bodyOf(code)
  console.log(wat)
  const buffer = watCompile(wat)
  const mod = new WebAssembly.Module(buffer)
  return new WebAssembly.Instance(mod, {})
}

export const benchWat = (fnName: string, params: string, iterations: number) =>
  `;;wasm
  (func $bench (export "bench")
    (local $index i32)

    ;; do
    (loop $loop
      (drop
        (call $${fnName} ${params})
      )

      ;; $index++
      (local.set $index (i32.add (local.get $index) (i32.const 1)))

      ;; if ($index !== iterations) continue $loop
      (br_if $loop (i32.ne (local.get $index) (i32.const ${iterations})))
    )
  )
`

const parseArgNumber = (x: string | number) => {
  if (typeof x === 'number') return x
  const n = parseFloat(x)
  if (n.toString() !== x) return x
  return n
}

const options = decarg(new Options())!

try {
  if (options.iterations) {
    bench(options)
  } else {
    console.log(run(options))
  }
} catch (error) {
  if (error) console.error(error)
  process.exit(1)
}
