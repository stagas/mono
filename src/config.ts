import { Struct } from './compiler'

const KB = 1024
const MB = 1024 * KB

export const PAGE_BYTES = 64 * KB
export const MEM_PADDING = 1 * PAGE_BYTES
const MB_PER_CHANNEL = 5
export const CHANNEL_BYTES = MB * MB_PER_CHANNEL
export const EVENTS_SIZE = 128
export const EVENTS = 4 * EVENTS_SIZE * Int32Array.BYTES_PER_ELEMENT

export const config = {
  /** Number of initial channels (default: `1`) */
  channels: 1,
  /** Maximum number of channels (default: `6`) */
  maxChannels: 6,
  /** Block size in number of elements (default: `128`) */
  blockSize: 128,
  /** Sample rate */
  sampleRate: 44100,
  /** Number of sample buffers. */
  sampleCount: 6,
  /** Sample channels (mono=1 stereo=2). */
  sampleChannels: 2,
  /** Sample duration. */
  sampleSeconds: 4,
  /** Sample pointers. */
  samplePointers: [],

  eventsPointer: MEM_PADDING,
  eventsSize: EVENTS_SIZE,
}

export const sampleBufferSizes = (() => {
  const channel = config.sampleRate * config.sampleSeconds
    + Object.keys(Struct.Buffer).length - 1
  const one = channel * config.sampleChannels
  const bytes = one * config.sampleCount * Float32Array.BYTES_PER_ELEMENT
  const pages = bytes / PAGE_BYTES
  return { one, channel, bytes, pages }
})()

export const memory = {
  initial: Math.ceil(
    (MEM_PADDING
      + EVENTS
      + sampleBufferSizes.bytes
      + CHANNEL_BYTES
      + config.channels * CHANNEL_BYTES) / PAGE_BYTES
  ),
  maximum: Math.ceil(
    (MEM_PADDING
      + EVENTS
      + sampleBufferSizes.bytes
      + CHANNEL_BYTES
      + config.maxChannels * CHANNEL_BYTES) / PAGE_BYTES
  ),
}

export const memPadding = (MEM_PADDING + EVENTS + sampleBufferSizes.bytes)

// create sample pointers
const { sampleCount, sampleChannels } = config
const sizes = sampleBufferSizes
const startPos = MEM_PADDING + EVENTS
export const samplePointers = [] as number[][]
for (let i = 0; i < sampleCount; i++) {
  const ptrs = [] as number[]
  samplePointers.push(ptrs)
  for (let c = 0; c < sampleChannels; c++) {
    const ptr = startPos + (
      (i * sizes.one + c * sizes.channel)
      * Float32Array.BYTES_PER_ELEMENT
    )
    ptrs.push(ptr)
  }
}
