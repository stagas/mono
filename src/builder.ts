import rfdc from 'rfdc'
import make, * as wat from 'wat-compiler'
import { Imports, Module, compile } from './compiler'
import * as lib from './lib.wat'
import { parse } from './parser'

// eslint-disable-next-line @typescript-eslint/no-explicit-any
let modlib: any
const copy = rfdc({ proto: true, circles: false })
const imports: Imports = {}

export const build = (s: string, scope = {}, extraLib = {}, extraRun = (_mod: Module) => []) => {
  console.time('build')

  if (!modlib) {
    console.time('lib')
    const libsrc = Object.values(lib).join('\n') + '\n' + Object.values(extraLib).join('\n')
    modlib = wat.compile(wat.parse(wat.tokenize('(module ' + libsrc + ')')))
    for (const code of modlib.module.codes) {
      const [params, result] = modlib.module.types[code.type_idx].split(',')
      imports[code.name] = { params: params.split(' '), result: result.split(' ')[0] }
    }
    console.timeEnd('lib')
  }

  // console.time('parse')
  const ast = parse(s)
  // console.timeEnd('parse')

  // console.time('compile')
  const mod = compile(ast, scope, imports)
  // console.timeEnd('compile')

  // console.time('make')
  const buffer = make(mod.toString(extraRun(mod)), { metrics: false }, copy(modlib))
  // console.timeEnd('make')

  console.timeEnd('build')

  return {
    module: mod,
    buffer,
  }
}
