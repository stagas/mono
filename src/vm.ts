import { Token } from 'lexer-next'
import { Arg } from './compiler'
import { Module } from './compiler'
import * as libmono from './lib/mono'
import * as libvm from './lib/vm'
import * as libwat from './lib/wat'
import { Linker, LinkerConfig } from './linker'
import { Type } from './typed'

const defaultConfig: LinkerConfig = {
  memory: {
    initial: 32, // TODO: autocalculate page size
    maximum: 32,
  },
  metrics: false,
}

const parseArg =
  /(?<arg>\.?\s*(?<id>[a-zA-Z_$][a-zA-Z0-9_$]*)(?<range>\[[^\]]*?\])?(=.*?(?<default>[^\s,)]+)?)?).*?(,|\)\s*=)/s

export class MonoParam {
  id!: Token
  sourceIndex!: number
  source!: {
    arg: string
    id: string
    range: string
    default: string
  }
  name!: string
  minValue!: number
  maxValue!: number
  defaultValue!: number
  normalValue!: number
  scaleValue!: number

  constructor(data: Partial<MonoParam>) {
    Object.assign(this, data)
    this.scaleValue = this.maxValue - this.minValue
    this.normalValue = this.normalize(this.defaultValue)
  }

  normalize(value: number) {
    return (value - this.minValue) / this.scaleValue
  }

  scale(normal: number) {
    return normal * this.scaleValue + this.minValue
  }
}

export class VM {
  isReady = false
  skipMono = false

  code = ''

  config: LinkerConfig

  linker: Linker
  module?: Module
  instance?: WebAssembly.Instance

  inputs: Float32Array[]
  outputs: Float32Array[]

  constructor(config: Partial<LinkerConfig> = {}) {
    this.config = { ...defaultConfig, ...config }
    this.linker = new Linker(this.config)
    this.inputs = [new Float32Array(this.linker.memory.buffer, 0x40000, 0x10000)] // TODO: config size
    this.outputs = [new Float32Array(this.linker.memory.buffer, 0, 0x10000)]
    this.linkLib()
  }

  linkLib() {
    this.linker.linkLib({ env: libvm.env, ...libwat })
    if (!this.skipMono) this.linker.linkMono(libmono)
    this.linker.linkSExpr(() => [libvm.fill({ type: this.f_type, params: this.f_params })])
  }

  get exports() {
    return this.instance?.exports as {
      [k: string]: ((...args: number[]) => number) | WebAssembly.Global | any
    } & libvm.VM
  }

  get funcs() {
    return this.linker.module!.funcs
  }

  get params(): MonoParam[] {
    return Object.values(this.module?.funcs ?? {})
      .map(f => [f.id, f.params.filter(p => p.export)])
      .filter(x => x[1].length)
      .map(
        ([id, params]: any) =>
          params.map((p: Arg) => {
            const export_id = p.id.as('export/' + id + '/' + p.id) as Token
            const export_id_min = p.id.as('export/' + id + '/' + p.id + '/min') as Token
            const export_id_max = p.id.as('export/' + id + '/' + p.id + '/max') as Token

            const minValue = (this.instance!.exports[export_id_min] as any)?.value ?? 0
            const maxValue = (this.instance!.exports[export_id_max] as any)?.value ?? 1
            const defaultValue = (this.instance!.exports[export_id] as any)?.value
              ?? (maxValue - minValue) * 0.5 + minValue

            // find the index of the dot `.` that made the param export by moving
            // backwards from the id index that we have access to here
            const code = p.id.source.input
            const sourceIndex = code.lastIndexOf('.', export_id.source.index)
            const source = code.slice(sourceIndex).match(parseArg)!.groups as MonoParam['source']

            return new MonoParam({
              id: export_id,
              sourceIndex,
              source,
              name: export_id.toString(),
              minValue,
              maxValue,
              defaultValue,
            })
          })
        // argToMonoParam({ ...p, default: p.originalDefault, id: p.id.as('export/' + id + '/' + p.id) as Token })
        // )
      )
      .flat()
  }

  get f_type(): Type {
    return this.linker.module!.typeOf(this.funcs.f?.body)
  }

  get f_params(): Arg[] {
    return this.funcs.f?.params ?? [] // .map(argToMonoParam)
  }

  async setCode(code: string) {
    this.linker.link(code)
    await this.linker.make()

    // update code only after linker has been successful
    this.code = code

    // we copy these because they are instantiated at two async steps
    // before the linker is ready, and "module" and "instance" can diverge
    // if we were to access them directly from the linker, leading to errors
    this.module = this.linker.module
    this.instance = this.linker.instance

    this.isReady = true
    this.config.metrics && console.log('mono: ready')
  }
}
