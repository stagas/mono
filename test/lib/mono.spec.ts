import { VM } from '../../src/vm'

const b = async (s: string) => {
  const vm = new VM()
  await vm.setCode(s)
  floats = new Float32Array(vm.linker.memory.buffer, 0, 5)
  return vm.exports
}

let floats: Float32Array

describe('mono', () => {
  it('lp', async () => {
    ;(await b('f(x)=lp(sin(t*x*pi2), 333.0, 2.5)')).fill(0, 5, 666)

    expect([...floats]).toMatchSnapshot()

    const bq = { x1: 0, x2: 0, y1: 0, y2: 0 }
    expect(
      Array.from({ length: 5 })
        .fill(0)
        .map((_, i) => lp(bq, Math.sin((i / 44100) * 666 * Math.PI * 2), 333, 2.5))
    ).toMatchSnapshot()
  })
})

const pi2 = 6.2831854820251465
const sampleRate = 44100

const lp = (b: any, x0: any, freq = 1000, Q = 1) => {
  const { y1, y2, x1, x2 } = b,
    w0 = (pi2 * freq) / sampleRate,
    sin_w0 = Math.sin(w0),
    cos_w0 = Math.cos(w0),
    alpha = sin_w0 / (2.0 * Q),
    b0 = (1.0 - cos_w0) / 2.0,
    b1 = 1.0 - cos_w0,
    b2 = b0,
    a0 = 1.0 + alpha,
    a1 = -2.0 * cos_w0,
    a2 = 1.0 - alpha,
    y0 = (x0 * b0 + x1 * b1 + x2 * b2 - y1 * a1 - y2 * a2) / a0
  b.y1 = y0
  b.y2 = y1
  b.x1 = x0
  b.x2 = x1

  return y0
}
