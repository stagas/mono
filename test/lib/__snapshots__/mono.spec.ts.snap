// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mono lp 1`] = `
Array [
  0,
  0.00005280714321997948,
  0.0003152576391585171,
  0.0009916839189827442,
  0.002277880208566785,
]
`;

exports[`mono lp 2`] = `
Array [
  0,
  0.00005280720298351749,
  0.00031525799979380265,
  0.0009916850854711705,
  0.0022778830220130373,
]
`;
